tests/test_serialization.py

```diff
@@ -0,0 +1,144 @@
+from web_poet.serialization import (
+    SerializedLeafData,
+    deserialize,
+    deserialize_leaf,
+    read_serialized_data,
+    register_serialization,
+    serialize,
+    serialize_leaf,
+    write_serialized_data,
+)
+
+
+def _assert_webpages_equal(p1: WebPage, p2: WebPage) -> None:
+    assert p1.response.body == p2.response.body
+    assert p1.response.status == p2.response.status
+    assert p1.response.headers == p2.response.headers
+    assert p1.response._encoding == p2.response._encoding
+    _assert_urls_equal(p1.response.url, p2.response.url)
+
+
+def _assert_urls_equal(u1: _Url, u2: _Url) -> None:
+    assert type(u1) == type(u2)
+    assert str(u1) == str(u2)
+
+
+def test_serialization_leaf() -> None:
+    data = {"a": "b", "c": 42}
+    serialized_data = serialize_leaf(data)
+    assert isinstance(serialized_data["json"], bytes)
+    assert json.loads(serialized_data["json"]) == data
+    deserialized_data = deserialize_leaf(dict, serialized_data)
+    assert data == deserialized_data
+
+
+def test_serialization_leaf_unsup() -> None:
+    class A:
+        pass
+
+    with pytest.raises(NotImplementedError):
+        serialize_leaf(A())
+
+    with pytest.raises(NotImplementedError):
+        deserialize_leaf(A, {})
+
+
+def test_serialization(book_list_html_response) -> None:
+    @attrs.define
+    class MyWebPage(WebPage):
+        url: ResponseUrl
+
+    url_str = "http://books.toscrape.com/index.html"
+    url = ResponseUrl(url_str)
+
+    serialized_deps = serialize([book_list_html_response, url])
+    assert serialized_deps["web_poet.page_inputs.http.HttpResponse"][
+        "body.html"
+    ] == bytes(book_list_html_response.body)
+    other_data = json.loads(
+        serialized_deps["web_poet.page_inputs.http.HttpResponse"]["other.json"]
+    )
+    assert other_data["url"] == url_str
+    assert (
+        serialized_deps["web_poet.page_inputs.url.ResponseUrl"]["txt"]
+        == url_str.encode()
+    )
+
+    po = MyWebPage(book_list_html_response, url)
+    deserialized_po = deserialize(MyWebPage, serialized_deps)
+    _assert_webpages_equal(po, deserialized_po)
+
+
+def test_serialization_httpresponse_encoding(book_list_html) -> None:
+    body = HttpResponseBody(bytes(book_list_html, "utf-8"))
+    resp_enc = HttpResponse(
+        url=ResponseUrl("http://books.toscrape.com/index.html"),
+        body=body,
+        encoding="utf-8",
+    )
+    assert resp_enc._encoding == "utf-8"
+    deserialized_resp_enc = deserialize_leaf(HttpResponse, serialize_leaf(resp_enc))
+    assert deserialized_resp_enc._encoding == "utf-8"
+
+    resp_noenc = HttpResponse(
+        url=ResponseUrl("http://books.toscrape.com/index.html"), body=body
+    )
+    assert resp_noenc._encoding is None
+    deserialized_resp_noenc = deserialize_leaf(HttpResponse, serialize_leaf(resp_noenc))
+    assert deserialized_resp_noenc._encoding is None
+
+
+def test_custom_functions() -> None:
+    class C:
```

Comment:
is fully qualified name works ok in this case?

Reply:
`serialize_leaf` doesn't use that, class names are only used to refer to serialized leaves in the `SerializedData` dict.